Experimental: torch.vulkan.compile_shader

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #154926
* #154925
* #154924
* #154678
* #154677
* __->__ #154676
* #154675

This adds an analogue to torch.mps.compile_shader for Vulkan. In order to get it off the ground, I went with our existing calling convention for Vulkan shaders, which is outlined in a comment.

Example usage:

1) build with USE_VULKAN=1

2) Create test_vulkan_compiler_shader.py:
```python
import torch
import torch.vulkan

t = torch.tensor([1., 2., -3.], device="vulkan")
abs_shader = torch.vulkan.compile_shader("myabs", open("path/to/absshader.glsl").read())
t_abs = torch.empty_like(t)
abs_shader(t_abs, t)
print(t_abs.cpu())
```

3) Create absshader.glsl (adapted from aten/src/ATen/native/vulkan/glsl/abs.glsl to account for us currently not doing the same preprocessing that gen_vulkan_spv.py does):
```

layout(std430) buffer;

/* Qualifiers: layout - storage - precision - memory */

layout(set = 0, binding = 0, FORMAT) uniform PRECISION restrict writeonly image3D   uOutput;
layout(set = 0, binding = 1)         uniform PRECISION                    sampler3D uInput;
layout(set = 0, binding = 2)         uniform PRECISION restrict           Block {
  ivec3 size;
} uBlock;

layout(local_size_x_id = 0, local_size_y_id = 1, local_size_z_id = 2) in;

void main() {
  const ivec3 pos = ivec3(gl_GlobalInvocationID);

  if (all(lessThan(pos, uBlock.size.xyz))) {
    const vec4 intex = texelFetch(uInput, pos, 0);
    imageStore(
        uOutput,
        pos,
        abs(intex));
  }
}
```

4) Run test_vulkan_compiler_shader.py.

Known issues (some of the solutions are in future PRs in the stack):
- [ ] We shell out to glslc (which we just blithely assume is installed and on PATH) when IIUC we should be linking with shaderc
- [X] Shader calling convention requires images; there is no way to use SSBOs instead (and no way that I'm aware of to create SSBO-backed Tensors from Python anyway)
- [ ] Shaders are required to use specialization constants for work group size (this seems like it might be OK?)
- [x] Shaders are required to use our "block" approach for passing the input size rather than, say, push constants
- [X] Shaders cannot receive any other arguments about the input, such as strides
- [X] Python APIs need to be underscore-prefixed to make them private


Pull-Request-resolved: https://github.com/pytorch/pytorch/pull/153991

swolchok committed:
Experimental: torch.vulkan.compile_shader

This adds an analogue to torch.mps.compile_shader for Vulkan. In order to get it off the ground, I went with our existing calling convention for Vulkan shaders, which is outlined in a comment.

Example usage:

1) build with USE_VULKAN=1

2) Create test_vulkan_compiler_shader.py:
```
import torch
import torch.vulkan

t = torch.tensor([1., 2., -3.], device="vulkan")
abs_shader = torch.vulkan.compile_shader("myabs", open("path/to/absshader.glsl").read())
t_abs = torch.empty_like(t)
abs_shader(t_abs, t)
print(t_abs.cpu())
```

3) Create absshader.glsl (adapted from aten/src/ATen/native/vulkan/glsl/abs.glsl to account for us currently not doing the same preprocessing that gen_vulkan_spv.py does):
```

layout(std430) buffer;

/* Qualifiers: layout - storage - precision - memory */

layout(set = 0, binding = 0, FORMAT) uniform PRECISION restrict writeonly image3D   uOutput;
layout(set = 0, binding = 1)         uniform PRECISION                    sampler3D uInput;
layout(set = 0, binding = 2)         uniform PRECISION restrict           Block {
  ivec3 size;
} uBlock;

layout(local_size_x_id = 0, local_size_y_id = 1, local_size_z_id = 2) in;

void main() {
  const ivec3 pos = ivec3(gl_GlobalInvocationID);

  if (all(lessThan(pos, uBlock.size.xyz))) {
    const vec4 intex = texelFetch(uInput, pos, 0);
    imageStore(
        uOutput,
        pos,
        abs(intex));
  }
}
```

4) Run test_vulkan_compiler_shader.py.

Known issues:
- [ ] We shell out to glslc (which we just blithely assume is installed and on PATH) when IIUC we should be linking with shaderc
- [ ] Shader calling convention requires images; there is no way to use SSBOs instead (and no way that I'm aware of to create SSBO-backed Tensors from Python anyway)
- [ ] Shaders are required to use specialization constants for work group size
- [ ] Shaders are required to use our "block" approach for passing the input size rather than, say, push constants
- [ ] Shaders cannot receive any other arguments about the input, such as strides
- [ ] `submit_compute_job_with_args_container` is largely copy/pasted

ghstack-source-id: d82123da8aee10b1c25b4d4258a46923ec17297c
Pull-Request-resolved: https://github.com/pytorch/pytorch/pull/153991

[ghstack-poisoned]

swolchok commented:
this was previously sent as #153991 and still has some unaddressed comments. will port over soonish so they don't go stale

swolchok committed:
Update on "Experimental: torch.vulkan.compile_shader"


This adds an analogue to torch.mps.compile_shader for Vulkan. In order to get it off the ground, I went with our existing calling convention for Vulkan shaders, which is outlined in a comment.

Example usage:

1) build with USE_VULKAN=1

2) Create test_vulkan_compiler_shader.py:
```
import torch
import torch.vulkan

t = torch.tensor([1., 2., -3.], device="vulkan")
abs_shader = torch.vulkan.compile_shader("myabs", open("path/to/absshader.glsl").read())
t_abs = torch.empty_like(t)
abs_shader(t_abs, t)
print(t_abs.cpu())
```

3) Create absshader.glsl (adapted from aten/src/ATen/native/vulkan/glsl/abs.glsl to account for us currently not doing the same preprocessing that gen_vulkan_spv.py does):
```

layout(std430) buffer;

/* Qualifiers: layout - storage - precision - memory */

layout(set = 0, binding = 0, FORMAT) uniform PRECISION restrict writeonly image3D   uOutput;
layout(set = 0, binding = 1)         uniform PRECISION                    sampler3D uInput;
layout(set = 0, binding = 2)         uniform PRECISION restrict           Block {
  ivec3 size;
} uBlock;

layout(local_size_x_id = 0, local_size_y_id = 1, local_size_z_id = 2) in;

void main() {
  const ivec3 pos = ivec3(gl_GlobalInvocationID);

  if (all(lessThan(pos, uBlock.size.xyz))) {
    const vec4 intex = texelFetch(uInput, pos, 0);
    imageStore(
        uOutput,
        pos,
        abs(intex));
  }
}
```

4) Run test_vulkan_compiler_shader.py.

Known issues:
- [ ] We shell out to glslc (which we just blithely assume is installed and on PATH) when IIUC we should be linking with shaderc
- [ ] Shader calling convention requires images; there is no way to use SSBOs instead (and no way that I'm aware of to create SSBO-backed Tensors from Python anyway)
- [ ] Shaders are required to use specialization constants for work group size
- [ ] Shaders are required to use our "block" approach for passing the input size rather than, say, push constants
- [ ] Shaders cannot receive any other arguments about the input, such as strides
- [ ] `submit_compute_job_with_args_container` is largely copy/pasted

Pull-Request-resolved: https://github.com/pytorch/pytorch/pull/153991

[ghstack-poisoned]

malfet reviewed (approved):
LGTM, provided green CI signal

swolchok committed:
Update on "Experimental: torch.vulkan.compile_shader"


This adds an analogue to torch.mps.compile_shader for Vulkan. In order to get it off the ground, I went with our existing calling convention for Vulkan shaders, which is outlined in a comment.

Example usage:

1) build with USE_VULKAN=1

2) Create test_vulkan_compiler_shader.py:
```python
import torch
import torch.vulkan

t = torch.tensor([1., 2., -3.], device="vulkan")
abs_shader = torch.vulkan.compile_shader("myabs", open("path/to/absshader.glsl").read())
t_abs = torch.empty_like(t)
abs_shader(t_abs, t)
print(t_abs.cpu())
```

3) Create absshader.glsl (adapted from aten/src/ATen/native/vulkan/glsl/abs.glsl to account for us currently not doing the same preprocessing that gen_vulkan_spv.py does):
```

layout(std430) buffer;

/* Qualifiers: layout - storage - precision - memory */

layout(set = 0, binding = 0, FORMAT) uniform PRECISION restrict writeonly image3D   uOutput;
layout(set = 0, binding = 1)         uniform PRECISION                    sampler3D uInput;
layout(set = 0, binding = 2)         uniform PRECISION restrict           Block {
  ivec3 size;
} uBlock;

layout(local_size_x_id = 0, local_size_y_id = 1, local_size_z_id = 2) in;

void main() {
  const ivec3 pos = ivec3(gl_GlobalInvocationID);

  if (all(lessThan(pos, uBlock.size.xyz))) {
    const vec4 intex = texelFetch(uInput, pos, 0);
    imageStore(
        uOutput,
        pos,
        abs(intex));
  }
}
```

4) Run test_vulkan_compiler_shader.py.

Known issues (some of the solutions are in future PRs in the stack):
- [ ] We shell out to glslc (which we just blithely assume is installed and on PATH) when IIUC we should be linking with shaderc
- [X] Shader calling convention requires images; there is no way to use SSBOs instead (and no way that I'm aware of to create SSBO-backed Tensors from Python anyway)
- [ ] Shaders are required to use specialization constants for work group size (this seems like it might be OK?)
- [x] Shaders are required to use our "block" approach for passing the input size rather than, say, push constants
- [X] Shaders cannot receive any other arguments about the input, such as strides
- [X] Python APIs need to be underscore-prefixed to make them private


Pull-Request-resolved: https://github.com/pytorch/pytorch/pull/153991

[ghstack-poisoned]

swolchok committed:
Update on "Experimental: torch.vulkan.compile_shader"


This adds an analogue to torch.mps.compile_shader for Vulkan. In order to get it off the ground, I went with our existing calling convention for Vulkan shaders, which is outlined in a comment.

Example usage:

1) build with USE_VULKAN=1

2) Create test_vulkan_compiler_shader.py:
```python
import torch
import torch.vulkan

t = torch.tensor([1., 2., -3.], device="vulkan")
abs_shader = torch.vulkan.compile_shader("myabs", open("path/to/absshader.glsl").read())
t_abs = torch.empty_like(t)
abs_shader(t_abs, t)
print(t_abs.cpu())
```

3) Create absshader.glsl (adapted from aten/src/ATen/native/vulkan/glsl/abs.glsl to account for us currently not doing the same preprocessing that gen_vulkan_spv.py does):
```

layout(std430) buffer;

/* Qualifiers: layout - storage - precision - memory */

layout(set = 0, binding = 0, FORMAT) uniform PRECISION restrict writeonly image3D   uOutput;
layout(set = 0, binding = 1)         uniform PRECISION                    sampler3D uInput;
layout(set = 0, binding = 2)         uniform PRECISION restrict           Block {
  ivec3 size;
} uBlock;

layout(local_size_x_id = 0, local_size_y_id = 1, local_size_z_id = 2) in;

void main() {
  const ivec3 pos = ivec3(gl_GlobalInvocationID);

  if (all(lessThan(pos, uBlock.size.xyz))) {
    const vec4 intex = texelFetch(uInput, pos, 0);
    imageStore(
        uOutput,
        pos,
        abs(intex));
  }
}
```

4) Run test_vulkan_compiler_shader.py.

Known issues (some of the solutions are in future PRs in the stack):
- [ ] We shell out to glslc (which we just blithely assume is installed and on PATH) when IIUC we should be linking with shaderc
- [X] Shader calling convention requires images; there is no way to use SSBOs instead (and no way that I'm aware of to create SSBO-backed Tensors from Python anyway)
- [ ] Shaders are required to use specialization constants for work group size (this seems like it might be OK?)
- [x] Shaders are required to use our "block" approach for passing the input size rather than, say, push constants
- [X] Shaders cannot receive any other arguments about the input, such as strides
- [X] Python APIs need to be underscore-prefixed to make them private


Pull-Request-resolved: https://github.com/pytorch/pytorch/pull/153991

[ghstack-poisoned]
